### ISSUE TITLE
Catch use of PhantomData that requires T: Sized

### ISSUE THREAD

**Issue by dtolnay** [rust-lang-nursery/rust-clippy#2308]
In Serde the traits we implement for PhantomData should have worked for T: ?Sized, but as a bug they required T: Sized -- fixed in https://github.com/serde-rs/serde/commit/4751627f1cd14cacdf216188ccbb9ab0831e2b3f. Could Clippy have caught this? I struggle to think of any use of PhantomData\<T> where T is an unconstrained type parameter where you would *not* want to allow ?Sized. As a starting point, maybe we could catch impls of the form:

```rust
impl<..., T, ...> Trait<...> for PhantomData<T> { /* ... */ }
```

where there are no trait bounds on T. In this case you pretty much always want T: ?Sized right?

**Comment by oli-obk** [comment]
The latter I agree with being a clear case where a `?Sized` bound is always the right thing.

I can't think of any counter examples for arbitrary bounds, we should probably just run it on a few big crates and see what happens.

**Comment by dtolnay** [comment]
I found a few instances of this in github search that would all want ?Sized:

- https://github.com/servo/heapsize/blob/442120a050de7ef69faf5f173260b30b7648d4fa/src/lib.rs#L278
- https://github.com/nox/ignominie/blob/59479077509fd14e4837e941539c026a9ccfdd31/src/lib.rs#L215
- https://github.com/hprose/hprose-rust/blob/b28ebcd69599c668eec1c1652ef4c538d13cff73/hprose/src/io/decoder.rs#L328
- https://github.com/asajeffrey/serde-roundtrip/blob/ffdcdcd01213945d98847416b0db739854262037/src/lib.rs#L332
- https://github.com/aprilwade/randomprime/blob/fb1e1200a8071ce5b4240b25aeb03463f2183b2a/reader_writer/src/primitive_types.rs#L239
- https://github.com/arcnmx/nue/blob/8aa158298828427d067bb84ad2769a3608420492/packed/src/lib.rs#L152
- https://github.com/lrs-lang/lib/blob/42b958a32161ec418bceffd340259b14274a3294/src/base/conv/from.rs#L79

[Here](https://github.com/serde-rs/serde/blob/v1.0.26/serde/src/de/mod.rs#L720) is one that would be a false positive 😿 -- using the T for an associated type that requires Sized.

**Comment by dtolnay** [comment]
Oh but the "false positive" has a where-clause `T: Deserialize<'de>` and we have `Deserialize<'de>: Sized`, so hopefully we can avoid recommending a ?Sized bound in that situation because it would do nothing.

**Comment by clarcharr** [comment]
Perhaps this lint could be generalised to requiring `Sized` in cases where it's not needed?
